repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java

```diff
@@ -0,0 +1,304 @@
+            // Update the same asset(entity)
+            throw new AtlasBaseException(OPERATION_NOT_SUPPORTED, "Can't update a specific version of contract");
+        }
+    }
+    private void processCreateContract(AtlasEntity entity, EntityMutationContext context) throws AtlasBaseException {
+        /*
+          Low-level Design
+               | Authorization
+               | Deserialization of the JSON
+               ---| Validation of spec
+               | Validation of contract
+               | Create Version
+               | Create Draft
+               ---| asset to contract sync
+               | Create Publish
+               ---| two-way sync of attribute
+         */
+
+        String contractQName = (String) entity.getAttribute(QUALIFIED_NAME);
+        validateAttribute(!contractQName.endsWith(String.format("/%s", CONTRACT_QUALIFIED_NAME_SUFFIX)), "Invalid qualifiedName for the contract.");
+
+        String contractString = (String) entity.getAttribute(ATTR_CONTRACT);
+        DataContract contract = DataContract.deserialize(contractString);
+        String datasetQName = contractQName.substring(0, contractQName.lastIndexOf('/'));
+        contractQName = String.format("%s/%s/%s", datasetQName, contract.getType().name(), CONTRACT_QUALIFIED_NAME_SUFFIX);
+        AtlasEntityWithExtInfo associatedAsset = getAssociatedAsset(datasetQName, contract.getType().name());
+
+        authorizeContractCreateOrUpdate(entity, associatedAsset);
+
+        boolean contractSync = syncContractCertificateStatus(entity, contract);
+        if (contractSync) {
+            contractString = DataContract.serialize(contract);
+            entity.setAttribute(ATTR_CONTRACT, contractString);
+        }
+
+        AtlasEntity currentVersionEntity = getCurrentVersion(associatedAsset.getEntity().getGuid());
+        Long newVersionNumber = 1L;
+        if (currentVersionEntity != null) {
+            // Contract already exist
+            Long currentVersionNumber = (Long) currentVersionEntity.getAttribute(ATTR_CONTRACT_VERSION);
+            List<String> attributes = getDiffAttributes(entity, currentVersionEntity);
+            if (attributes.isEmpty()) {
+                // No changes in the contract, Not creating new version
+                removeCreatingVertex(context, entity);
+                return;
+            } else if (isEqualContract(contractString, (String) currentVersionEntity.getAttribute(ATTR_CONTRACT))) {
+                resetAllRelationshipAttributes(entity);
+                // No change in contract, metadata changed
+                updateExistingVersion(context, entity, currentVersionEntity);
+                newVersionNumber = currentVersionNumber;
+            } else {
+                // contract changed (metadata might/not changed). Create new version.
+                newVersionNumber =  currentVersionNumber + 1;
+
+                resetAllRelationshipAttributes(entity);
```

Comment:
We needed to add reset logic here as well as we are adding all relationships to the dataset and these user-added relationships are messing up the system.